### PR TITLE
Show zone from player position & %pos%

### PR DESCRIPTION
**Pull request by SohKey**
Added %pos% to share position in chat:
image

Re-implemented %item% to share item in hand:

image

code is probably ugly (i'm starting in nodejs) please give me advises :)

### CODE REVIEW

**Review by Sceat — CHANGES_REQUESTED** [review]
Thanks for the contribution! don't forget to run `npm run format` before pushing commits :)

There are some problems with the approach you used, i left a few comments and also rewrited the file, you can see in a comment below a cleaner implementation

**Comment by Sceat** [comment]
```js
import nbt from 'prismarine-nbt'
import minecraftData from 'minecraft-data'

import { item_to_slot } from '../items.js'
import logger from '../logger.js'
import execute_command from '../commands/commands.js'
import { VERSION } from '../settings.js'
import { Formats, world_chat_msg } from '../chat.js'
import { World } from '../events.js'
import items from '../../data/items.json' assert { type: 'json' }

const mcData = minecraftData(VERSION)
const log = logger(import.meta)

function is_command_function(message) {
  return message.trimStart()[0] === '/'
}

function slot_to_chat({ nbtData, itemCount, itemId }) {
  const tag = nbt.simplify(nbtData)
  const { name } = mcData.items[itemId]
  const chat = {
    ...JSON.parse(tag.display.Name),
    hoverEvent: {
      action: 'show_item',
      contents: {
        id: name,
        itemCount,
        tag: JSON.stringify(tag),
      },
    },
  }
  return chat
}

export default {
  /** @type {import('../context.js').Observer} */
  observe({ client, get_state, world, dispatch }) {
    const chat_mapper = {
      '%item%': () => {
        const { held_slot_index, inventory } = get_state()
        const item = inventory[held_slot_index + 36] // For the player 0 is the first item in hotbar. But for the game the hotbat begin at 36.
        if (item) {
          const { type, count } = item
          return slot_to_chat(item_to_slot(items[type], count))
        }
      },
      '%item\\d%': word => {
        const slot_number = parseInt(word.match(/\d/)[0]) + 36 // same
        const { inventory } = get_state()
        const item = inventory[slot_number]
        if (item) {
          const { type, count } = item
          return slot_to_chat(item_to_slot(items[type], count))
        }
      },
      '%pos%': () => {
        const {
          position: { x, y },
        } = get_state()
        return { text: `[${x}, ${y}] `, ...Formats.INFO }
      },
    }

    client.on('chat', packet => {
      const { message } = packet
      const { nickname } = get_state()

      if (is_command_function(message)) {
        log.info({ sender: client.uuid, command: message }, 'Command')
        execute_command({ world, message, sender: client, get_state, dispatch })
        return
      } else {
        log.info({ sender: client.uuid, message }, 'Message')
      }

      world_chat_msg({
        world,
        message: [
          { text: `${nickname}: `, ...Formats.BASE, italic: false },
          ...message.split(' ').map(word => {
            for (const pattern in chat_mapper)
              if (word.match(pattern)) return chat_mapper[pattern](word)
            return { text: `${word} `, ...Formats.BASE, italic: false }
          }),
        ],
        client,
      })
    })
    const on_chat = options => client.write('chat', options)
    const on_private_message = ({ receiver_username, options }) => {
      if (receiver_username === client.username) client.write('chat', options)
    }

    world.events.on(World.CHAT, on_chat)
    world.events.on(World.PRIVATE_MESSAGE, on_private_message)

    client.once('end', () => {
      world.events.off(World.CHAT, on_chat)
      world.events.off(World.PRIVATE_MESSAGE, on_private_message)
    })
  },
}
```

File: src/player/chat.js [hunk 1]
```diff
@@ -34,6 +34,17 @@ function slot_to_chat({ nbtData, itemCount, itemId }) {
   return chat
 }
 
+function share_pos(x,y) {
+  const chat = {
+    translate:"chat.type.text",
+    text:"Position", color: 'gold',
```
[the remark this answers is not in the document]
Comment (SohKey) [reply src/player/chat.js:40]: I wanted to share the zone and on hover the coordonates appear : 
image

File: src/player/chat.js [hunk 2]
```diff
@@ -48,32 +59,115 @@ export default {
...
+            client,
+          })
+        }
+
+        else if (message.match(/(%pos%)/)) {
+          const pos_link = message.split(/(%pos%)/).map(part => {
+            if (part.match(/(%pos%)/)) {
+              const { position } = get_state()
+              var posX = position["x"]
+              var posY = position["y"]
```
[the remark this answers is not in the document]
Comment (SohKey) [reply src/player/chat.js:96]: sorry ;( 
Im aware now x)

**Comment by SohKey** [comment]
> Thanks for the contribution! don't forget to run `npm run format` before pushing commits :)
> 
> There are some problems with the approach you used (like having multiple pos and item in the same msg), i left a few comments and also **rewrited the file to show you a leaner implementation** _(unable to use `replaceAll` because of the `\d`)_
> 
> ```js
> import nbt from 'prismarine-nbt'
> import minecraftData from 'minecraft-data'
> 
> import { item_to_slot } from '../items.js'
> import logger from '../logger.js'
> import execute_command from '../commands/commands.js'
> import { VERSION } from '../settings.js'
> import { Formats, world_chat_msg } from '../chat.js'
> import { World } from '../events.js'
> import items from '../../data/items.json' assert { type: 'json' }
> 
> const mcData = minecraftData(VERSION)
> const log = logger(import.meta)
> 
> function is_command_function(message) {
>   return message.trimStart()[0] === '/'
> }
> 
> function slot_to_chat({ nbtData, itemCount, itemId }) {
>   const tag = nbt.simplify(nbtData)
>   const { name } = mcData.items[itemId]
>   const chat = {
>     ...JSON.parse(tag.display.Name),
>     hoverEvent: {
>       action: 'show_item',
>       contents: {
>         id: name,
>         itemCount,
>         tag: JSON.stringify(tag),
>       },
>     },
>   }
>   return chat
> }
> 
> export default {
>   /** @type {import('../context.js').Observer} */
>   observe({ client, get_state, world, dispatch }) {
>     const chat_mapper = {
>       '%item%': () => {
>         const { held_slot_index, inventory } = get_state()
>         const item = inventory[held_slot_index + 36] // For the player 0 is the first item in hotbar. But for the game the hotbat begin at 36.
>         if (item) {
>           const { type, count } = item
>           return slot_to_chat(item_to_slot(items[type], count))
>         }
>       },
>       '%item\\d%': word => {
>         const slot_number = parseInt(word.match(/\d/)[0]) + 36 // same
>         const { inventory } = get_state()
>         const item = inventory[slot_number]
>         if (item) {
>           const { type, count } = item
>           return slot_to_chat(item_to_slot(items[type], count))
>         }
>       },
>       '%pos%': () => {
>         const {
>           position: { x, y },
>         } = get_state()
>         return { text: `[${x}, ${y}] `, ...Formats.INFO }
>       },
>     }
> 
>     client.on('chat', packet => {
>       const { message } = packet
>       const { nickname } = get_state()
> 
>       if (is_command_function(message)) {
>         log.info({ sender: client.uuid, command: message }, 'Command')
>         execute_command({ world, message, sender: client, get_state, dispatch })
>         return
>       } else {
>         log.info({ sender: client.uuid, message }, 'Message')
>       }
> 
>       world_chat_msg({
>         world,
>         message: [
>           { text: `${nickname}: `, ...Formats.BASE, italic: false },
>           ...message.split(' ').map(word => {
>             for (const pattern in chat_mapper)
>               if (word.match(pattern)) return chat_mapper[pattern](word)
>             return { text: `${word} `, ...Formats.BASE, italic: false }
>           }),
>         ],
>         client,
>       })
>     })
>     const on_chat = options => client.write('chat', options)
>     const on_private_message = ({ receiver_username, options }) => {
>       if (receiver_username === client.username) client.write('chat', options)
>     }
> 
>     world.events.on(World.CHAT, on_chat)
>     world.events.on(World.PRIVATE_MESSAGE, on_private_message)
> 
>     client.once('end', () => {
>       world.events.off(World.CHAT, on_chat)
>       world.events.off(World.PRIVATE_MESSAGE, on_private_message)
>     })
>   },
> }
> ```

**Comment by SohKey** [comment]
> Thanks for the contribution! don't forget to run `npm run format` before pushing commits :)
> 
> There are some problems with the approach you used (like having multiple pos and item in the same msg), i left a few comments and also **rewrited the file to show you a leaner implementation** _(unable to use `replaceAll` because of the `\d`)_
> 
> ```js
> import nbt from 'prismarine-nbt'
> import minecraftData from 'minecraft-data'
> 
> import { item_to_slot } from '../items.js'
> import logger from '../logger.js'
> import execute_command from '../commands/commands.js'
> import { VERSION } from '../settings.js'
> import { Formats, world_chat_msg } from '../chat.js'
> import { World } from '../events.js'
> import items from '../../data/items.json' assert { type: 'json' }
> 
> const mcData = minecraftData(VERSION)
> const log = logger(import.meta)
> 
> function is_command_function(message) {
>   return message.trimStart()[0] === '/'
> }
> 
> function slot_to_chat({ nbtData, itemCount, itemId }) {
>   const tag = nbt.simplify(nbtData)
>   const { name } = mcData.items[itemId]
>   const chat = {
>     ...JSON.parse(tag.display.Name),
>     hoverEvent: {
>       action: 'show_item',
>       contents: {
>         id: name,
>         itemCount,
>         tag: JSON.stringify(tag),
>       },
>     },
>   }
>   return chat
> }
> 
> export default {
>   /** @type {import('../context.js').Observer} */
>   observe({ client, get_state, world, dispatch }) {
>     const chat_mapper = {
>       '%item%': () => {
>         const { held_slot_index, inventory } = get_state()
>         const item = inventory[held_slot_index + 36] // For the player 0 is the first item in hotbar. But for the game the hotbat begin at 36.
>         if (item) {
>           const { type, count } = item
>           return slot_to_chat(item_to_slot(items[type], count))
>         }
>       },
>       '%item\\d%': word => {
>         const slot_number = parseInt(word.match(/\d/)[0]) + 36 // same
>         const { inventory } = get_state()
>         const item = inventory[slot_number]
>         if (item) {
>           const { type, count } = item
>           return slot_to_chat(item_to_slot(items[type], count))
>         }
>       },
>       '%pos%': () => {
>         const {
>           position: { x, y },
>         } = get_state()
>         return { text: `[${x}, ${y}] `, ...Formats.INFO }
>       },
>     }
> 
>     client.on('chat', packet => {
>       const { message } = packet
>       const { nickname } = get_state()
> 
>       if (is_command_function(message)) {
>         log.info({ sender: client.uuid, command: message }, 'Command')
>         execute_command({ world, message, sender: client, get_state, dispatch })
>         return
>       } else {
>         log.info({ sender: client.uuid, message }, 'Message')
>       }
> 
>       world_chat_msg({
>         world,
>         message: [
>           { text: `${nickname}: `, ...Formats.BASE, italic: false },
>           ...message.split(' ').map(word => {
>             for (const pattern in chat_mapper)
>               if (word.match(pattern)) return chat_mapper[pattern](word)
>             return { text: `${word} `, ...Formats.BASE, italic: false }
>           }),
>         ],
>         client,
>       })
>     })
>     const on_chat = options => client.write('chat', options)
>     const on_private_message = ({ receiver_username, options }) => {
>       if (receiver_username === client.username) client.write('chat', options)
>     }
> 
>     world.events.on(World.CHAT, on_chat)
>     world.events.on(World.PRIVATE_MESSAGE, on_private_message)
> 
>     client.once('end', () => {
>       world.events.off(World.CHAT, on_chat)
>       world.events.off(World.PRIVATE_MESSAGE, on_private_message)
>     })
>   },
> }
> ```

File: src/player/chat.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (Sceat) [reply src/player/chat.js]: maybe a `%zone%` then

File: src/player/action_bar.js [hunk 3]
```diff
@@ -54,12 +57,14 @@ export default {
   observe({ client, get_state, world, events, signal }) {
     aiter(abortable(setInterval(2000, null, { signal }))).forEach(() => {
       const state = get_state()
+      const { position } = get_state()
```
Comment (Sceat) [inline src/player/action_bar.js:60]: get_state doesn't need to be called 2 times

File: src/player/action_bar.js [hunk 4]
```diff
@@ -76,13 +81,15 @@ export default {
         const max_health_changed = last_max_health !== max_health
         const stats_points_changed =
           last_remaining_stats_point !== remaining_stats_point
+        const { position } = get_state()
```
Comment (Sceat) [inline src/player/action_bar.js:84]: the state is already present line 76, `get_state` should only be used when there is no other way to access it

File: src/player/teleportation_stones.js [hunk 5]
```diff
@@ -63,6 +63,32 @@ function teleportation_stones_in_chunk(world, chunk_x, chunk_z) {
...
 }
 
+ function is_player_near(Xplayer,Yplayer, Xstone, Ystone) {
+  const radius = 30
+  if((Math.abs(Xstone-Xplayer) <= radius) && (Math.abs(Ystone-Yplayer) <= radius)) {
+    return true
+  }else{
+    return false
+  }
+}
```
Comment (Sceat) [inline src/player/teleportation_stones.js:73]: there already are usable distance function in `src/math.js`

File: src/player/teleportation_stones.js [hunk 6]
```diff
@@ -63,6 +63,32 @@ function teleportation_stones_in_chunk(world, chunk_x, chunk_z) {
...
+ */
+export function closest_stone(world,player_pos_x,player_pos_y) {
+  const stones = world.teleportation_stones
+  const zone = stones
+  .filter(stone => is_player_near(player_pos_x, player_pos_y, stone.position.x, stone.position.y))
+  if(zone.length === 0) {
+    return "wilderness"
+  }
+  return `${zone.map(stone => stone.name)}`
+  }
```
Comment (Sceat) [inline src/player/teleportation_stones.js:90]: That function is ambiguous as it says closest_stone not closest_stones_names, it should return a single stone or have its name changed. But it's better to return a simple stone. (also wilderness is not a stone)

You don't need the world, just the stones array, also it's better to use an object argument, or in the case a signature as below

> A player can't be in 2 different zones btw

> the distance function comes from `src/math.js`

```suggestion
const PLAYER_ZONE_DISTANCE = 30

// note that a player at equal distance of 2 stones
// would be at the zone limit so it doesn't matter if we say one or the other
export function closest_stone(stones, { position }) {
  return stones.find(stone => distance(stone.position, position) <= PLAYER_ZONE_DISTANCE)
}
```

that function returns a stone or `undefined` so you can verify as
`const zone = closest_stone() ?? 'wilderness'`

File: src/player/chat.js [hunk 7]
```diff
@@ -34,11 +35,50 @@ function slot_to_chat({ nbtData, itemCount, itemId }) {
   return chat
 }
 
+function position_to_chat(zone,x,y) {
+  const chat = {
+    translate:"chat.type.text",
+    text:`§6§nPosition:§r§a ${zone}`,
```
Comment (Sceat) [inline src/player/chat.js:41]: you have to use text components instead of mojang's magic char `§`

File: src/player/chat.js [hunk 8]
```diff
@@ -34,11 +35,50 @@ function slot_to_chat({ nbtData, itemCount, itemId }) {
   return chat
 }
 
+function position_to_chat(zone,x,y) {
+  const chat = {
+    translate:"chat.type.text",
```
Comment (Sceat) [inline src/player/chat.js:40]: This is unneeded

File: src/player/chat.js [hunk 9]
```diff
@@ -34,11 +35,50 @@ function slot_to_chat({ nbtData, itemCount, itemId }) {
...
+    text:`§6§nPosition:§r§a ${zone}`,
+    hoverEvent: {
+      action:"show_text",
+      value:`§dX: ${x} Y: ${y}`,
```
Comment (Sceat) [inline src/player/chat.js:44]: We should not show the exact position when a player simply wants to show the zone

**Review by Sceat — CHANGES_REQUESTED** [review]

File: src/player/chat.js [hunk 10]
```diff
@@ -37,8 +37,43 @@ function slot_to_chat({ nbtData, itemCount, itemId }) {
...
+      '%pos%': () => {
+        const { position } = get_state()
+        const closest_zone =
+          closest_stone(world, position)?.name ?? 'Wilderness'
+        const chat = {
+          text: `Position: ${closest_zone}`,
+          color: 'gold',
+          hoverEvent: {
+            action: 'show_text',
+            value: `X: ${Math.round(position.x)} Y: ${Math.round(position.y)}`,
```
Comment (DeltaEvo) [inline src/player/chat.js:67]: I think here we want to show `X and Z` instead as `Y` is the height

File: src/player/chat.js [hunk 11]
```diff
@@ -48,30 +83,16 @@ export default {
...
-          ],
-        },
+        message: [
+          { text: `${nickname}: `, ...Formats.BASE, italic: false },
+          ...message.split(' ').map(word => {
+            for (const pattern in chat_mapper)
+              if (word.match(pattern)) return chat_mapper[pattern](word)
+            return { text: `${word} `, ...Formats.BASE, italic: false }
+          }),
+        ],
```
Comment (DeltaEvo) [inline src/player/chat.js:95]: I would prefer for us to keep the translating key (cc @Sceat) as this allows to not hardcode the format here and use the one that we put in the ressource pack

```suggestion
        message: {
          translate: 'chat.type.text',
          with: [
            {
              text: client.username,
            },
            message.split(' ').map(word => {
              for (const pattern in chat_mapper)
                if (word.match(pattern)) return chat_mapper[pattern](word)
              return { text: `${word} `, ...Formats.BASE, italic: false }
            }),
          ],
        },
```

File: src/player/chat.js [hunk 12]
```diff
@@ -48,30 +83,16 @@ export default {
...
-            },
-            formatted_message,
-          ],
-        },
+        message: [
+          { text: `${nickname}: `, ...Formats.BASE, italic: false },
+          ...message.split(' ').map(word => {
+            for (const pattern in chat_mapper)
+              if (word.match(pattern)) return chat_mapper[pattern](word)
+            return { text: `${word} `, ...Formats.BASE, italic: false }
```
Comment (DeltaEvo) [inline src/player/chat.js:93]: I think this implementation doesn't work as intended as no space is added if the word match, so I'm expecting for `Hello %item% World` to remove the space after the item.
Could you test that @SohKey

Also I'm not a big fan of the extra trailing space at the end

(cc @Sceat)

**Comment by DeltaEvo** [comment]
@Sceat could you resolve the comments that you think are fixed ?

File: src/player/chat.js (hunk 10 above)
Comment (SohKey) [reply src/player/chat.js:67]: Yes, i just checked and for me z was the height, i will change that

File: src/player/chat.js (hunk 12 above)
Comment (SohKey) [reply src/player/chat.js:93]: yep, you are right the bug is as you say

**Review by SohKey — COMMENTED** [review]
- Correction for %pos% y --> z
- Added suggered code
- Correction for player without item in hand trying to link item in chat
   -> Issue: The actual code dont allow to send a msg to prevent player
- Issue: player has to put space between %item% to link various items (%item1%%item2% return only %item1%)

File: src/player/chat.js (hunk 11 above)
Comment (Sceat) [reply src/player/chat.js]: oh this is resource handled ? interesting

File: src/player/chat.js (hunk 12 above)
Comment (Sceat) [reply src/player/chat.js]: a space should be added to `slot_to_chat`

**Comment by DeltaEvo** [comment]
I'm not sure to understand why it is not supposed to write that :thinking:

**Comment by Sceat** [comment]
> I'm not sure to understand why it is not supposed to write that 🤔

%item1% %item2% is valid

**Comment by DeltaEvo** [comment]
Yes and why `%item1%%item2%` is not ?

File: src/player/chat.js [hunk 13]
```diff
@@ -69,7 +92,15 @@ export default {
...
-            formatted_message,
+            message.split(' ').map(word => {
+              for (const pattern in chat_mapper)
+                if (
+                  word.match(pattern) &&
+                  chat_mapper[pattern](word) !== undefined
+                )
+                  return chat_mapper[pattern](word)
+              return { text: `${word}`, ...Formats.BASE, italic: false }
+            }),
```
Comment (DeltaEvo) [inline src/player/chat.js:103]:
```suggestion
            message.split(new RegExp(`(${Object.keys(chat_mapper).join('|')})`).map(word => {
              for (const pattern in chat_mapper)
                if (
                  word.match(pattern) &&
                  chat_mapper[pattern](word) !== undefined
                )
                  return chat_mapper[pattern](word)
              return { text: word }
            }),
```

File: src/player/chat.js [hunk 14]
```diff
@@ -37,6 +37,42 @@ function slot_to_chat({ nbtData, itemCount, itemId }) {
...
+      '%item%': () => {
+        const { held_slot_index, inventory } = get_state()
+        const item = inventory[held_slot_index + 36] // For the player 0 is the first item in hotbar. But for the game the hotbat begin at 36.
+        if (item !== undefined) {
+          const { type, count } = item
+          return slot_to_chat(item_to_slot(items[type], count))
+        }
+        return undefined
+      },
+      '%item\\d%': word => {
```
Comment (DeltaEvo) [inline src/player/chat.js:50]:
```suggestion
      [/%item\d%/.source]: word => {
```

**Review by DeltaEvo — APPROVED** [review]

**Review by DeltaEvo — APPROVED** [review]

**Pull request merged**